File: elixir/knapsack/lib/knapsack.ex
defmodule Knapsack do
  @doc """
  Return the maximum value that a knapsack can carry.
  """
  @spec maximum_value(items :: [%{value: integer, weight: integer}], maximum_weight :: integer) ::
          integer
  def maximum_value(items, maximum_weight)
  def maximum_value([], _), do: 0
  def maximum_value([%{weight: w, value: _} | _], maximum_weight) when w > maximum_weight, do: 0
  def maximum_value([%{weight: w, value: v} | rest], maximum_weight) do
    do_not_take = maximum_value(rest, maximum_weight)
    take = v + maximum_value(rest, maximum_weight - w)
    max(do_not_take, take)
  end

end

File: c/knapsack/knapsack.c
#include "knapsack.h"
#include <stdlib.h>

int maximum_value(int maximum_weight, item_t *items, size_t item_count) {
  if (item_count == 0) {
    return 0;
  }
  int *data = calloc(maximum_weight + 1, sizeof(int));
  for (size_t i = 0; i < item_count; ++i) {
    int weight = items[i].weight;
    for (int j = maximum_weight; j >= weight; --j) {
      int take = data[j - weight] + items[i].value;
      int dont_take = data[j];
      data[j] = dont_take > take ? dont_take : take;
    }
  }
  int result = data[maximum_weight];
  free(data);
  return result;
}
